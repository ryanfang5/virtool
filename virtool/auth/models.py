from sqlalchemy import Column, Enum, String

from virtool.pg.base import Base
from virtool.pg.utils import SQLEnum


class ResourceType(str, SQLEnum):
    """
    Types of resources a permission can apply to (e.g. 'app', 'sample', 'group').

    """

    app = "app"
    group = "group"


class Action(str, SQLEnum):
    """
    Conserved actions that can be performed on a resource and
    are controlled by permissions.
    """

    create = "create"
    update = "update"
    delete = "delete"
    modify = "modify"
    remove = "remove"


class SQLPermission(Base):
    """
    SQL model to store permissions

    """

    __tablename__ = "permissions"

    id = Column(String, primary_key=True)
    name = Column(String)
    description = Column(String)
    resource_type = Column(Enum(ResourceType))
    action = Column(Enum(Action))
