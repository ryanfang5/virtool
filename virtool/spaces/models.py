from sqlalchemy import (
    Column,
    DateTime,
    Integer,
    String,
)

from virtool.pg.base import Base


class SpaceModel(Base):
    __tablename__ = "space"

    id = Column(Integer, primary_key=True)
    name = Column(String, unique=True)
    description = Column(String)
    created_at = Column(DateTime, nullable=False)
    updated_at = Column(DateTime, nullable=False)
    created_by = Column(String, nullable=False)

