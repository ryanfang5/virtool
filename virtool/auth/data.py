from typing import List

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncEngine, AsyncSession

from virtool.auth.models import SQLPermission
from virtool_core.models.auth import PermissionMinimal


class AuthData:
    def __init__(self, pg: AsyncEngine):
        self._pg = pg

    async def find(self, resource_type) -> List[PermissionMinimal]:
        """
        List all possible permissions.

        :return: a list of all permissions

        """
        statement = select(SQLPermission)

        if resource_type:
            statement = statement.filter_by(resource_type=resource_type)

        async with AsyncSession(self._pg) as session:
            result = await session.execute(statement)

        return [
            PermissionMinimal(**permission.to_dict()) for permission in result.scalars()
        ]
