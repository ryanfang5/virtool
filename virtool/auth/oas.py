from pydantic import BaseModel


class ListPermissionsResponse(BaseModel):
    class Config:
        schema_extra = {
            "example": [
                {
                    "id": 1,
                    "name": "create_sample",
                    "description": "Required for creating a sample",
                    "resource_type": "app",
                    "action": "create",
                },
                {
                    "id": 2,
                    "name": "modify_subtraction",
                    "description": "Required for modifying a subtraction",
                    "resource_type": "app",
                    "action": "modify",
                },
            ]
        }
