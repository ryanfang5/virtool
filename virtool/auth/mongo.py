from typing import List

from virtool.auth.permissions import PermissionType
from virtool.data.errors import ResourceNotFoundError
from virtool.mongo.core import DB
from virtool.mongo.utils import get_one_field


async def check_in_mongo(mongo: DB, user_id: str, permission: PermissionType) -> bool:
    """
    Check permissions in Mongo.
    """

    return (
        await mongo.users.count_documents(
            {"_id": user_id, f"permissions.{permission.name}": True}, limit=1
        )
        == 1
    )


async def list_permissions_in_mongo(mongo: DB, user_id: str) -> List[PermissionType]:
    """
    List user permissions in Mongo.
    """

    permissions = await get_one_field(mongo.users, "permissions", user_id)

    if permissions:
        return sorted(
            [permission for permission in permissions if permissions[permission]]
        )

    raise ResourceNotFoundError()
