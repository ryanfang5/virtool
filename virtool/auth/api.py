from typing import Optional

from aiohttp_pydantic import PydanticView
from aiohttp_pydantic.oas.typing import r200

from virtool.api.response import json_response
from virtool.auth.models import ResourceType
from virtool.auth.oas import ListPermissionsResponse
from virtool.data.utils import get_data_from_req
from virtool.http.routes import Routes

routes = Routes()


@routes.view("/source/permissions")
class SourceView(PydanticView):
    async def get(
        self, resource_type: Optional[ResourceType] = None
    ) -> r200[ListPermissionsResponse]:
        """
        List permissions.

        Lists all Virtool permissions.

        The list can be filtered by resource type using the
        `resource_type` query parameter.

        Status Codes:
            200: Successful operation
        """

        return json_response(
            await get_data_from_req(self.request).auth.find(resource_type)
        )
