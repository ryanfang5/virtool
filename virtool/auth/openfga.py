from typing import List, Union

from openfga_sdk import (
    ApiClient,
    CheckRequest,
    TupleKey,
    ReadRequest,
    WriteRequest,
    TupleKeys, ApiException,
)
from openfga_sdk.api import open_fga_api
from virtool_core.models.enums import Permission

from virtool.users.utils import generate_base_permissions


async def check_in_open_fga(
    open_fga: ApiClient,
    user_id: str,
    permission: Permission,
    object_type: str,
    object_id: Union[str, int],
) -> bool:
    """
    Check a permission in OpenFGA.
    """
    api_instance = open_fga_api.OpenFgaApi(open_fga)

    body = CheckRequest(
        tuple_key=TupleKey(
            user=f"user:{user_id}",
            relation=permission.name,
            object=f"{object_type}:{object_id}",
        ),
    )

    response = await api_instance.check(body)

    return response.allowed


async def list_permissions_in_open_fga(
    open_fga: ApiClient, user_id: str, object_type: str, object_id: Union[str, int]
) -> dict:
    """
    List permissions for a user in OpenFGA.
    """

    api_instance = open_fga_api.OpenFgaApi(open_fga)

    permissions = generate_base_permissions()

    body = ReadRequest(
        tuple_key=TupleKey(user=f"user:{user_id}", relation="member", object="group:"),
    )

    response = await api_instance.read(body)

    groups = [relation_tuple.key.object for relation_tuple in response.tuples]

    for group in groups:
        body = ReadRequest(
            tuple_key=TupleKey(
                user=f"{group}#member", object=f"{object_type}:{object_id}"
            ),
        )

        response = await api_instance.read(body)

        for relation_tuple in response.tuples:
            permissions.update({relation_tuple.key.relation: True})

    body = ReadRequest(
        tuple_key=TupleKey(user=f"user:{user_id}", object=f"{object_type}:{object_id}")
    )

    response = await api_instance.read(body)

    for relation_tuple in response.tuples:
        permissions.update({relation_tuple.key.relation: True})

    return permissions


async def add_group_permissions_in_open_fga(
    open_fga: ApiClient, group_id: str, permissions: List[Permission], object_type: str, object_id: Union[str, int]
):
    """
    Add permissions to a group in OpenFGA.
    """

    api_instance = open_fga_api.OpenFgaApi(open_fga)

    for permission in permissions:

        body = WriteRequest(
                writes=TupleKeys(
                    tuple_keys=[TupleKey(
                        user=f"group:{group_id}#member",
                        relation=permission.name,
                        object=f"{object_type}:{object_id}",
                    ),
                    ]
                )
            )
        try:
            await api_instance.write(body)
        except ApiException:
            pass


async def add_user_permissions_in_open_fga(
    open_fga: ApiClient, user_id: str, permissions: List[Permission], object_type: str, object_id: Union[str, int]
):
    """
    Add permissions for a user in OpenFGA.
    """

    api_instance = open_fga_api.OpenFgaApi(open_fga)

    for permission in permissions:

        body = WriteRequest(
                writes=TupleKeys(
                    tuple_keys=[TupleKey(
                        user=f"user:{user_id}",
                        relation=permission.name,
                        object=f"{object_type}:{object_id}",
                    ),
                    ]
                )
            )
        try:
            await api_instance.write(body)
        except ApiException:
            pass


async def remove_group_permissions_in_open_fga(
    open_fga: ApiClient,
    group_id: str,
    permissions: List[Permission],
    object_type: str,
    object_id: Union[str, id],
):
    """
    Remove permissions from a group in OpenFGA.
    """

    api_instance = open_fga_api.OpenFgaApi(open_fga)

    for permission in permissions:
        body = WriteRequest(
            deletes=TupleKeys(
                tuple_keys=[
                    TupleKey(
                        user=f"group:{group_id}#member",
                        relation=permission.name,
                        object=f"{object_type}:{object_id}",
                    ),
                ],
            ),
        )
        try:
            await api_instance.write(body)
        except ApiException:
            pass


async def remove_user_permissions_in_open_fga(
    open_fga: ApiClient,
    user_id: str,
    permissions: List[Permission],
    object_type: str,
    object_id: Union[str, id],
):
    """
    Remove permissions for a user in OpenFGA.
    """
    api_instance = open_fga_api.OpenFgaApi(open_fga)

    for permission in permissions:
        body = WriteRequest(
            deletes=TupleKeys(
                tuple_keys=[
                    TupleKey(
                        user=f"user:{user_id}",
                        relation=permission.name,
                        object=f"{object_type}:{object_id}",
                    ),
                ],
            ),
        )
        try:
            await api_instance.write(body)
        except ApiException:
            pass
