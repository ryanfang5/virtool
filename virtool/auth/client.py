import asyncio
from abc import ABC, abstractmethod
from typing import Union

from openfga_sdk import TupleKey, OpenFgaApi
from virtool_core.models.enums import Permission

from virtool.auth.mongo import (
    check_in_mongo,
    list_permissions_in_mongo,
)
from virtool.auth.openfga import (
    check_in_open_fga,
    list_permissions_in_open_fga,
    add_in_open_fga,
    remove_in_open_fga,
)
from virtool.auth.relationships import BaseRelationship
from virtool.data.errors import ResourceNotFoundError

from virtool.mongo.core import DB


class AbstractAuthorizationClient(ABC):
    @abstractmethod
    async def check(
        self,
        user_id: str,
        permission: Permission,
        object_type: str,
        object_id: Union[str, int],
    ) -> bool:
        ...

    @abstractmethod
    async def list_permissions(
        self, user_id: str, object_type: str, object_id: Union[str, int]
    ) -> dict:
        ...

    @abstractmethod
    async def add(self, relationship: BaseRelationship):
        ...

    @abstractmethod
    async def remove(self, relationship: BaseRelationship):
        ...


class AuthorizationClient(AbstractAuthorizationClient):
    def __init__(self, mongo: DB, open_fga: OpenFgaApi):
        self.mongo = mongo
        self.open_fga = open_fga

    async def check(
        self,
        user_id: str,
        permission: Permission,
        object_type: str,
        object_id: Union[str, int],
    ) -> bool:
        """
        Check whether a user has a permission on a resource.
        """

        mongo_result, open_fga_result = await asyncio.gather(
            *[
                check_in_mongo(self.mongo, user_id, permission),
                check_in_open_fga(
                    self.open_fga, user_id, permission, object_type, object_id
                ),
            ]
        )

        return mongo_result or open_fga_result

    async def list_permissions(
        self, user_id: str, object_type: str, object_id: Union[str, int]
    ) -> dict:
        """
        List permissions for a user.
        """
        try:
            return await list_permissions_in_mongo(self.mongo, user_id)
        except ResourceNotFoundError:
            return await list_permissions_in_open_fga(
                self.open_fga, user_id, object_type, object_id
            )

    async def add(self, relationship: BaseRelationship):
        """
        Add an authorization relationship.
        """
        try:
            await relationship.add(self.mongo)
        except ResourceNotFoundError:
            pass

        tuple_list = [
            TupleKey(
                user=f"{relationship.user_type}:{relationship.user_id}",
                relation=relation,
                object=f"{relationship.object_type}:{relationship.object_name}",
            )
            for relation in relationship.relation
        ]

        await add_in_open_fga(self.open_fga, tuple_list)

    async def remove(self, relationship: BaseRelationship):
        """
        Remove an authorization relationship.
        """
        try:
            await relationship.remove(self.mongo)
        except ResourceNotFoundError:
            pass

        tuple_list = [
            TupleKey(
                user=f"{relationship.user_type}:{relationship.user_id}",
                relation=relation,
                object=f"{relationship.object_type}:{relationship.object_name}",
            )
            for relation in relationship.relation
        ]

        await remove_in_open_fga(self.open_fga, tuple_list)
