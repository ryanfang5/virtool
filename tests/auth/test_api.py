import pytest
from sqlalchemy.ext.asyncio import AsyncSession
from virtool_core.models.enums import Permission

from virtool.auth.models import SQLPermission, ResourceType, Action


@pytest.mark.apitest
@pytest.mark.parametrize(
    "resource,status",
    [(ResourceType.app, 200), (ResourceType.group, 200), ("invalid", 400), (None, 200)],
)
async def test_find(spawn_client, resource, status, pg, snapshot):
    client = await spawn_client(authorize=True)

    async with AsyncSession(pg) as session:
        session.add_all(
            [
                SQLPermission(
                    id=Permission.create_sample,
                    name="Create Sample",
                    resource_type=ResourceType.app,
                    action=Action.create,
                    description="Required for creating a sample",
                ),
                SQLPermission(
                    id=Permission.modify_subtraction,
                    name="Modify Subtraction",
                    resource_type=ResourceType.app,
                    action=Action.modify,
                    description="Required for modifying a subtraction",
                ),
            ]
        )
        await session.commit()

    url = "/source/permissions"

    if resource:
        url += f"?resource_type={resource}"

    resp = await client.get(url)

    assert resp.status == status
    assert await resp.json() == snapshot
