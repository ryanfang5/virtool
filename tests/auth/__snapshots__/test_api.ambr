# name: test_find[uvloop-None-200]
  <class 'list'> [
    <class 'dict'> {
      'action': 'create',
      'description': 'Required for creating a sample',
      'id': 'create_sample',
      'name': 'Create Sample',
      'resource_type': 'app',
    },
    <class 'dict'> {
      'action': 'modify',
      'description': 'Required for modifying a subtraction',
      'id': 'modify_subtraction',
      'name': 'Modify Subtraction',
      'resource_type': 'app',
    },
  ]
---
# name: test_find[uvloop-app-200]
  <class 'list'> [
    <class 'dict'> {
      'action': 'create',
      'description': 'Required for creating a sample',
      'id': 'create_sample',
      'name': 'Create Sample',
      'resource_type': 'app',
    },
    <class 'dict'> {
      'action': 'modify',
      'description': 'Required for modifying a subtraction',
      'id': 'modify_subtraction',
      'name': 'Modify Subtraction',
      'resource_type': 'app',
    },
  ]
---
# name: test_find[uvloop-group-200]
  <class 'list'> [
  ]
---
# name: test_find[uvloop-invalid-400]
  <class 'list'> [
    <class 'dict'> {
      'ctx': <class 'dict'> {
        'enum_values': <class 'list'> [
          'app',
          'group',
        ],
      },
      'in': 'query string',
      'loc': <class 'list'> [
        'resource_type',
      ],
      'msg': "value is not a valid enumeration member; permitted: 'app', 'group'",
      'type': 'type_error.enum',
    },
  ]
---
